### PR TITLE
Add public interface to lockfile interospection

### LINKED ISSUE #1725: crate_universe: exposing the bazel lockfile format
Hello. I am making a [gazelle](https://github.com/bazelbuild/bazel-gazelle) plugin for rust for automatic build file management: https://github.com/Calsign/gazelle_rust

One of the things that the gazelle plugin handles is dependencies on external crates pulled in through crate_universe. This requires parsing lockfiles. Depending on the configuration, rules_rust currently uses both cargo lockfiles (`Cargo.lock`) and custom "bazel lockfiles" (`Cargo.Bazel.lock`). While it is not currently possible to go fully cargo-free (i.e., no `Cargo.lock`), [this comment](https://github.com/bazelbuild/rules_rust/issues/1674#issuecomment-1330900533) suggests that allowing a cargo-free setup is desired. Thus to support cargo-free mode, the gazelle plugin needs to be able to parse bazel lockfiles.

Parsing bazel lockfiles is tricky because the interface is not exposed and it is subject to change across rules_rust versions. My current approach involves [depending on cargo_bazel](https://github.com/Calsign/gazelle_rust/blob/main/rust_parser/BUILD.bazel#L22) from the version of rules_rust used in each project and [patching rules_rust](https://github.com/Calsign/gazelle_rust/blob/main/patches/rules_rust.patch) to expose the required modules from cargo_bazel (`context` and `utils`). This is suboptimal.

I would like the gazelle plugin to be usable without needing to patch rules_rust. What's the best approach here? I'm thinking the `context` and `utils` modules could be split out into a separate crate that is publicly exposed. (This would also make the gazelle plugin build faster.) Is that reasonable? If so, I'm happy to put up the PR, just wanted to open a discussion first.

Separately it could make sense to integrate the gazelle plugin into rules_rust if that is desired. But I'll open a different issue for that.

### PR DESCRIPTION
**Pull request by illicitonion**
https://github.com/Calsign/gazelle_rust currently relies on a patch to rules_rust which makes the `Context` struct public.

Instead, supply the information they need via a public API.

All names (and whether we want to hide these structs behind traits and just return `impl Trait`s) are open to discussion.

Fixes #1725

cc @Calsign

### CODE REVIEW

**Comment by Calsign** [comment]
Thanks! This looks great. I haven't gotten around to doing this yet so thanks for picking up the slack.

Just for reference, there are additional changes which could make this situation even better:
1. Split the logic for parsing lockfiles into a separate crate so that we don't need to build all of cargo_bazel to build gazelle_rust
2. Add some kind of test checking compatibility of the public lockfile API

For 1, I'll do this at some point in the future if I'm sufficiently motivated. For now it's not that big of a deal.

For 2, that would need some kind of guarantee from rules_rust which the project is understandably not ready to provide. Improving the status quo to not require a patch is great and definitely good enough for now. But at some point in the future it would be nice to have a stronger guarantee.

File: crate_universe/src/lockfile/public.rs
Comment (UebelAndre) [inline crate_universe/src/lockfile/public.rs:47]: If this is gonna be a public interface can we get full docstrings on these?

File: crate_universe/src/lockfile/public.rs
Comment (UebelAndre) [inline crate_universe/src/lockfile/public.rs:49]: Sounds like a good idea?

File: crate_universe/src/lib.rs
Comment (UebelAndre) [inline crate_universe/src/lib.rs:7]: Maybe we should have an `api` module instead? I worry about leaking things from `lockfile` since now we'd need to explicitly state `pub(crate)` Maybe that's a good thing though? I'll defer to you 😄

**Review by UebelAndre — APPROVED** [review]
This seems fine to me!

Had a few nits/questions but nothing urgent or worth blocking merge 😄

**Comment by illicitonion** [comment]
I think this is ready to go (and agree with @Calsign's future extensions, but would treat them as extensions :)) - the first three commits are just https://github.com/bazelbuild/rules_rust/pull/2525

File: crate_universe/test_data/cargo_bazel_lockfile/multi_package-cargo-bazel-lock.json
Comment (UebelAndre) [inline crate_universe/test_data/cargo_bazel_lockfile/multi_package-cargo-bazel-lock.json:1]: @illicitonion question: how was this file generated? Could you add some docs for it?

File: crate_universe/src/api/lockfile.rs
Comment (UebelAndre) [inline crate_universe/src/api/lockfile.rs:125]: @illicitonion What is this testing?

Comment (illicitonion) [reply crate_universe/src/api/lockfile.rs:125]: See https://github.com/bazelbuild/rules_rust/pull/2637 :)

File: crate_universe/test_data/cargo_bazel_lockfile/multi_package-cargo-bazel-lock.json
Comment (illicitonion) [reply crate_universe/test_data/cargo_bazel_lockfile/multi_package-cargo-bazel-lock.json:1]: See https://github.com/bazelbuild/rules_rust/pull/2637 :)

**Pull request merged**
